### PR TITLE
Adding Patch Endpoint for Resources

### PR DESCRIPTION
**Pull request by GavinMendelGleason**
This endpoint will apply a patch to a resource

### CODE REVIEW

File: src/server/routes.pl [hunk 1]
```diff
@@ -2488,6 +2492,36 @@
...
+            _{ patch : Patch } :< JSON),
+        error(bad_api_document(JSON, [patch]), _)),
+    % We should take options about final state here.
+    api_report_errors(
+        patch,
+        Request,
+        (   (   memberchk(search(Search), Request)
+            ->  true
+            ;   Search = []),
+            param_value_search_or_json_required(Search, JSON, author, text, Author),
```
Comment (matko) [inline src/server/routes.pl:2512]: There is a weird inconsistency here between endpoints.
Some endpoints seem to expect commit info to be in a special commit_info subdocument in the payload. But this (and I presume other places where this was copied from) would expect author and message to be toplevel.

File: src/core/api/api_patch.pl [hunk 2]
```diff
@@ -12,16 +13,69 @@
...
+        (
+            patch_id_pairs(Patch, Patch_And_Ids),
+            mapm({Context, Merged_Options}/[Patch-_,Conflict,Ids]>>(
+                     apply_diff_ids_captures(Context, Patch, Conflict, Ids, Merged_Options)
+                 ),
+                 Patch_And_Ids,
+                 Conflicts,
+                 Ids_List,
+                 Empty,
+                 _Captures
```
Comment (matko) [inline src/core/api/api_patch.pl:62]: if we thread through captures shouldn't we be checking to make sure that all captures are bound in the end?

Unless of course there is no way to actually use captures and this was just to make the calls happy.

File: src/core/api/api_patch.pl [hunk 3]
```diff
@@ -12,16 +13,69 @@
...
+                ),
+            % actually an unzip
+            exclude([null]>>true, Conflicts, Witnesses),
+            (   Witnesses = []
+            ->  true
+            ;   maplist([conflict(Conflict),Conflict]>>true,
```
Comment (matko) [inline src/core/api/api_patch.pl:68]: should this not be an include?
This will only succeed if all patch fragments had conflicts.

Comment (GavinMendelGleason) [reply src/core/api/api_patch.pl:68]: Yes, this is wrong.

File: src/core/api/api_patch.pl [hunk 4]
```diff
@@ -12,16 +13,69 @@
...
+                        Witnesses,
+                        Conflicts),
```
Comment (matko) [inline src/core/api/api_patch.pl:70]: this rebinds Conflicts, probably not great.

**Review by matko — CHANGES_REQUESTED** [review]

File: src/server/routes.pl (hunk 1 above)
Comment (GavinMendelGleason) [reply src/server/routes.pl:2512]: Let's fix that where the inconsistency arises.

**Review by matko — APPROVED** [review]

**Pull request merged**